Add initial support for shading rate control

bjoeris opened the pull request:
- Adds ability to detect and enable Vulkan support for Fragment Density Map (FDM) and Variable Rate Shading (VRS).
- Adds `ShadingRatePattern` for specifying the pixel density for each region of the render target.
- Adds support for attaching a `ShadingRatePattern` to a RenderPass.
- Adds application settings and knob flags to enable FDM or VRS.

*Note:* VRS support with the DX12 backend is not yet implemented. Attempting to enable FDM or VRS with DX12 will cause an error.

hysw commented on include/ppx/grfx/grfx_foveation.h:
```diff
@@ -0,0 +1,134 @@
+
+    uint32_t EncodeFragmentSizeFoveationTextureValue(uint8_t fragmentWidth, uint8_t fragmentHeight) const;
+    uint32_t EncodeFragmentDensityFoveationTextureValue(uint8_t xDensity, uint8_t yDensity) const;
+    void     FillFoveationTextureData(FoveationPatternType patternType, uint8_t* pData) const;
```
Add FDM/VRS to function names?

bjoeris replied:
There are private FDM/VRS-specific versions of these functions. These functions allow code to specify either fragment size or density, and it will be encoded correctly for whichever foveation mode is being used.

bjoeris replied:
I've changed the way the shading rate images are filled, and moved this to a follow-up PR.

hysw commented on include/ppx/grfx/grfx_enums.h:
```diff
@@ -259,6 +259,22 @@ enum Filter
+    FOVEATION_VRS  = 2,
+};
+
+enum FoveationPatternType
```
Also consider loading pattern from file, either consist of the foveation pattern data themselves or pixel per degree that we can interpolate&quantize to a pattern.

hysw replied:
SG. It's a suggestion and does not need to be in the PR.

hysw commented on src/ppx/grfx/grfx_foveation.cpp:
```diff
@@ -0,0 +1,335 @@
+}
+
+} // namespace grfx
+} // namespace ppx
```
nit: add newline at end of file.

bjoeris replied:
Done.

hysw commented on src/ppx/grfx/grfx_foveation.cpp:
```diff
@@ -0,0 +1,335 @@
+    // Create foveation texture
+    uint32_t texel_w = foveationCapabilites.fdm.texelSize.min.width;
+    uint32_t texel_h = foveationCapabilites.fdm.texelSize.min.height;
+    mTextureWidth    = pCreateInfo->fbWidth / texel_w;
```
Should this be `ceil` instead of `floor`?

bjoeris replied:
I think floor is correct because `texel_w` is the minimum width in the render target that can correspond to a single pixel in the FDM, so `fbWidth / texel_w` is the maximum allowed width of the FDM, so ceil would make the FDM too wide.

hysw replied:
Discussed in chat, according to [vkspec] and [Quacomm extension], texture width should be `((fbWidth + texel_w - 1)/texel_w)`

[vkspec]: https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#fragmentdensitymap-fetch-density-value
[Quacomm extension]: https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap12.html#VUID-VkImageCreateInfo-fragmentDensityMapOffset-06514

bjoeris replied:
This now uses ceil (src/ppx/grfx/vk/vk_shading_rate.cpp:34).

hysw commented on src/ppx/grfx/grfx_foveation.cpp:
```diff
@@ -0,0 +1,335 @@
+
+    uint32_t texel_w = foveationCapabilites.vrs.minTexelSize.x;
+    uint32_t texel_h = foveationCapabilites.vrs.minTexelSize.y;
+    mTextureWidth    = pCreateInfo->fbWidth / texel_w;
```
Should this be `ceil` instead of `floor`?

hysw commented on src/ppx/grfx/vk/vk_device.cpp:
```diff
@@ -243,6 +261,168 @@ Result Device::ConfigureFeatures(const grfx::DeviceCreateInfo* pCreateInfo, VkPh
+                    mapRateToSupported[encoded] = mapRateToSupported[shrinkWidth];
+                }
+                else {
+                    mapRateToSupported[encoded] = std::max<uint8_t>(mapRateToSupported[shrinkWidth], mapRateToSupported[shrinkHeight]);
```
nit: The max here is a bit confusing, this line almost always picks shrinkHeight due to encoding.

bjoeris replied:
I think the max is necessary. E.g. assume the supported shading rates are 1x1 (encoded=0), 2x2 (encoded=5), 4x4 (encoded=9). Then 4x2 should map to 2x2, but shrinkHeight is 4x1, which maps to 1x1.

hysw replied:
I'm not claiming it's unnecessary, but it can be confusing on the substitution criteria (therefore it's a nit).
E.g. with available rate 1x1, 2x1, 4x1, 1x2, 1x4, 2x2 the max strategy will pick 4x1 instead rather than 2x2 for 4x4, although I don't think any hardware have such weird set of rates.

bjoeris replied:
I added a comment clarifying the behaviour here. That will be in the follow-up PR that handles filling in the shading rate images.

hysw commented on include/ppx/grfx/grfx_foveation.h:
```diff
@@ -0,0 +1,134 @@
+
+} // namespace grfx
+} // namespace ppx
+#endif // ppx_grfx_foveation_h
```
nit: add newline at end of file.

apazylbe reviewed (commented):
Two high level comments:
- VRS and FDM use here seems to be very foveation specific, which itself is more XR specific. I am wondering if we should think about making some parts more generic where it can be used for other purposes
- Not sure about what DX12 support, but we will ideally want to add similar support for it in the future. Not sure if the current behaviour on DX12 is appropriate (i.e., reasonable errors)

hysw commented on include/ppx/grfx/grfx_foveation.h:
```diff
@@ -0,0 +1,134 @@
+        {
+            // List of supported encoded shading rates
+            uint32_t supportedRateCount = 0;
+            uint8_t  supportedRates[11];
```
The maximum encoded valid rate is 4x4, 0b1010 = 10, so array need 11 elements. Although imo it's probably better to have 16 = 2^4 elements.

hysw commented on include/ppx/grfx/grfx_foveation.h:
```diff
@@ -0,0 +1,134 @@
+        bool  supportPipelineVRS   = false;
+        bool  supportPrimitiveVRS  = false;
+        bool  supportAttachmentVRS = false;
+        uint2 minTexelSize;
+        uint2 maxTexelSize;
```
This makes me really want to add `{Extent,Offset}{2,3}D<T>`

bjoeris replied:
These have been switched to use a new Extent2D struct. It's not a template, but could easily be turned into a template when we have a use case with another type.

bjoeris commented:
> Hello! Curious, do you have a branch with any project using foveation? Could be something like 01_triangle, but could help understand how this would integrate into an app code 🙂

Yes, I have a fovation benchmark I put together to measure the performance impact of a bunch of foveation related variables, as well as a commit that enables foveation in 04_cube. https://github.com/bjoeris/bigwheels/tree/experimental-foveation

bjoeris commented on include/ppx/grfx/vk/vk_util.h:
```diff
@@ -80,6 +80,13 @@ VkImageAspectFlags DetermineAspectMask(VkFormat format);
 VmaMemoryUsage ToVmaMemoryUsage(grfx::MemoryUsage value);
 
+template <typename TVkStruct1, typename TVkStruct2>
+void InsertPNext(TVkStruct1& struct1, TVkStruct2& struct2)
```
Done.

bjoeris commented on include/ppx/application.h:
```diff
@@ -301,6 +301,12 @@ struct ApplicationSettings
+        // If `foveationMode` is not `FOVEATION_NONE`, then the pipeline
+        // targeting the application's RenderPass must use the same foveation
+        // mode (`GraphicsPipelineCreateInfo.foveationMode`).
+        grfx::FoveationMode        foveationMode        = grfx::FOVEATION_NONE;
```
After talking with Aliya, this has been removed to avoid over-complicating the Application class.

This means apps will need to create their own RenderPasses to use foveation. This might be simplified by adding hooks for applications to customize the Swapchain/RenderPasses created by Application.

bjoeris commented on include/ppx/grfx/grfx_foveation.h:
```diff
@@ -0,0 +1,134 @@
+namespace ppx {
+namespace grfx {
+
+//! @struct FoveationCapabilities
```
Done.

bjoeris commented on include/ppx/grfx/grfx_foveation.h:
```diff
@@ -0,0 +1,134 @@
+    struct
+    {
+        bool supported                                 = false;
+        bool supportsDynamicImageView                  = false;
+        bool supportsNonSubsampledImages               = false;
+        bool supportsAdditionalInvocations             = false;
+        bool supportsDeferred                          = false;
+        bool subsampledLoads                           = false;
+        bool subsampledCoarseReconstructionEarlyAccess = false;
```
These have been reduced to just supportsFDM, supportsPipelineVRS, supportsPrimitiveVRS, supportsAttachmentVRS. We can add more later as we actually use them.

bjoeris commented on include/ppx/grfx/grfx_device.h:
```diff
@@ -178,11 +182,13 @@ class Device
 
     grfx::QueuePtr GetAnyAvailableQueue() const;
 
+    const grfx::FoveationCapabilities& GetFoveationCapabilities() const { return mFoveationCapabilities; }
```
This has been renamed to `GetShadingRateCapabilities`.

bjoeris commented on src/ppx/grfx/vk/vk_render_pass.cpp:
```diff
@@ -277,27 +394,35 @@ VkResult CreateTransientRenderPass(
+    if (foveationFDM) {
+        InsertPNext(vkci, fdmInfo);
+    }
+    VkResult vkres = vk::CreateRenderPass(
```
The `VK_KHR_create_renderpass2` extension is now required in order to enable either VRS or FDM (src/ppx/grfx/vk/vk_device.cpp:165-183,329,356).

 `vk::RenderPass` now uses `vkCreateRenderPass` unless it has a `ShadingRatePattern` (src/ppx/grfx/vk/vk_render_pass.cpp:139) or `ShadingRateMode` (src/ppx/grfx/vk/vk_render_pass.cpp:329).

hysw reviewed (commented):
LGTM

apazylbe commented on include/ppx/grfx/vk/vk_device.h:
```diff
@@ -90,23 +91,37 @@ class Device
     Result ConfigureQueueInfo(const grfx::DeviceCreateInfo* pCreateInfo, std::vector<float>& queuePriorities, std::vector<VkDeviceQueueCreateInfo>& queueCreateInfos);
     Result ConfigureExtensions(const grfx::DeviceCreateInfo* pCreateInfo);
     Result ConfigureFeatures(const grfx::DeviceCreateInfo* pCreateInfo, VkPhysicalDeviceFeatures& features);
+    void   ConfigureShadingRate(
```
nit: ConfigureShadingRate**Capabilities**? and same for the two functions below

bjoeris replied:
These have been renamed.

apazylbe commented on src/ppx/grfx/grfx_shading_rate.cpp:
Do we need this file?

bjoeris replied:
Removed. It's used in a follow up PR, where it is added back.

apazylbe commented:
Just a quick note, contributing page (https://github.com/google/bigwheels/blob/main/CONTRIBUTING.md#for-developers-contributing-a-patch) mentions "Please add new commits to your branch, instead of amending your commit. Adding new commits makes it easier for the reviewer to see what has changed since the last review."
It's easier to see changes that way, particular for larger PRs

bjoeris commented:
> Just a quick note, contributing page (https://github.com/google/bigwheels/blob/main/CONTRIBUTING.md#for-developers-contributing-a-patch) mentions "Please add new commits to your branch, instead of amending your commit. Adding new commits makes it easier for the reviewer to see what has changed since the last review." It's easier to see changes that way, particular for larger PRs

Apologies, that makes sense. I'll do that for future PRs.

bjoeris commented on include/ppx/grfx/grfx_device.h:
```diff
@@ -20,6 +20,7 @@
 #include "ppx/grfx/grfx_command.h"
 #include "ppx/grfx/grfx_descriptor.h"
 #include "ppx/grfx/grfx_draw_pass.h"
+#include "ppx/grfx/grfx_shading_rate.h"
```
Removed.

bjoeris commented on include/ppx/grfx/grfx_helper.h:
```diff
@@ -213,16 +213,17 @@ struct ImageUsageFlags
+            bool depthStencilAttachment        : 1;
+            bool transientAttachment           : 1;
+            bool inputAttachment               : 1;
+            bool shadingRateImageNv            : 1;
```
Done.

bjoeris commented on include/ppx/grfx/grfx_shading_rate.h:
```diff
@@ -0,0 +1,119 @@
+
+// ShadingRateCapabilities
+//
+// Information about GPU support shading rate features.
```
Done.

bjoeris commented on src/ppx/grfx/grfx_draw_pass.cpp:
```diff
@@ -26,10 +26,11 @@ namespace internal {
+    this->width               = obj.width;
+    this->height              = obj.height;
+    this->renderTargetCount   = obj.renderTargetCount;
+    this->pShadingRatePattern = obj.pShadingRatePattern;
```
Added.

bjoeris commented on src/ppx/grfx/vk/vk_device.cpp:
```diff
@@ -265,6 +288,117 @@ Result Device::ConfigureFeatures(const grfx::DeviceCreateInfo* pCreateInfo, VkPh
+    InsertPNext(properties, fdmProperties);
+    mFnGetPhysicalDeviceProperties2(physicalDevice, &properties);
+
+    pShadingRateCapabilities->supportsFDM      = fdmFeatures.fragmentDensityMap == VK_TRUE;
```
Done.

bjoeris commented on src/ppx/grfx/vk/vk_render_pass.cpp:
```diff
@@ -120,32 +124,48 @@ Result RenderPass::CreateRenderPass(const grfx::internal::RenderPassCreateInfo*
     subpassDependencies.dependencyFlags     = 0;
 
-    VkRenderPassCreateInfo vkci = {VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO};
+    VkRenderPassCreateInfo vkci = {};
```
The sType has been put back in.

bjoeris commented on src/ppx/grfx/vk/vk_render_pass.cpp:
```diff
@@ -287,7 +314,7 @@ VkResult CreateTransientRenderPass(
     subpassDependencies.dependencyFlags     = 0;
 
-    VkRenderPassCreateInfo vkci = {VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO};
+    VkRenderPassCreateInfo vkci = {};
```
Done.

bjoeris commented on src/ppx/grfx/vk/vk_shading_rate.cpp:
```diff
@@ -0,0 +1,313 @@
+    mShadingRateMode         = pCreateInfo->shadingRateMode;
+    const auto& capabilities = GetDevice()->GetShadingRateCapabilities();
+
+    grfx::ImageCreateInfo imageCreateInfo;
```
Done.

bjoeris commented on src/ppx/grfx/vk/vk_shading_rate.cpp:
```diff
@@ -0,0 +1,313 @@
+            PPX_ASSERT_MSG(false, "Cannot create ShadingRatePattern for ShadingRateMode " << pCreateInfo->shadingRateMode);
+            return ppx::ERROR_FAILED;
+    }
+    if (pCreateInfo->texelSize.width == 0 && pCreateInfo->texelSize.height == 0) {
```
Done.

Keenuts reviewed (approved):
Wasn't able to test the cube or other samples on the branch on my devices (lacks extensions) and won't be able to on a device that has those before next week.
I'll just ask you to run those on a compatible device with validation enabled, and make sure we don't trigger any validation errors.

bjoeris commented on src/ppx/grfx/vk/vk_shading_rate.cpp:
```diff
@@ -0,0 +1,315 @@
+    vkci2.flags = vkci.flags;
+
+    mAttachments.clear();
+    mAttachments.reserve(vkci.attachmentCount + 1);
```
The `+1` is for the `VkAttachmentDescription2` added for the denisty map or vrs attachment.

Keenuts replied:
Right, in the other function, the `push_back`.
Right.
What I'm not a fan of is we have attributes which are initialized across multiple functions, but function are not isolated at all. If you remove this reserve in one function, it brings a nasty bug in another function when we push-back and potentially end-up with bad pointers.

In understand why you split it, as each override will have to initialize part of the structure differently, but there might be something better to do.
Something where the vector is passed around, and the scope where each element is known to be fully initialized is a bit clearer.

Maybe something like:
```cpp
mAttachments = GetBaseAttachments();
auto additionnal_attachments_it = AddImplementationAttachments(attachments);

// Calls the base function to initialize create info, then internally calls the implem specific function. But it's a const method.
// The passed iterator tells us where the implementation specific attachments are.
mCreateInfo = MakeCreateInfo(attachments, additionnal_attachments_it);
```

bjoeris replied:
>  If you remove this reserve in one function, it brings a nasty bug in another function when we push-back and potentially end-up with bad pointers.

I believe the `reserve()` call here is a very minor optimization, not critical. The `push_back()` call in `InitializeImpl()` might reallocate the vector memory, but the `pAttachments` will be updated to point to the new memory.

Since this makes the code harder to understand and provides very little benefit, I've removed the `reserve()` call.

> Something where the vector is passed around, and the scope where each element is known to be fully initialized is a bit clearer.

I don't think I fully understand the suggestion. The fact that `InitializeImpl()` modifies `mAttachments` is an implementation detail that I don't think needs to be exposed. All of the pointers are valid both before and after the `InitializeImpl()` call.

I did some refactoring/renaming/commenting that I hope makes this code more clear. If you still think the change you were suggesting makes sense, please clarify.

Keenuts replied:
Thanks for the refactoring!

apazylbe commented:
Thanks! Would you be able to update the PR description to more accurately reflect what is included in this PR? 
(I haven't taken a look after your last changes)

bjoeris commented:
> Wasn't able to test the cube or other samples on the branch on my devices (lacks extensions) and won't be able to on a device that has those before next week. I'll just ask you to run those on a compatible device with validation enabled, and make sure we don't trigger any validation errors.

Thanks for the reminder to check the validation layers. I did find a couple of errors and added commits to fix these.

One of the errors is going to be a bit annoying: we are only allowed to enable one of FDM or VRS when creating the device. This means `Application` needs to know which shading rate mode to pass to the device create info. For now I just added this to `ApplicationSettings`, but I'll also need a way to set this from a knob.

bjoeris commented on src/ppx/grfx/vk/vk_render_pass.cpp:
```diff
@@ -217,15 +240,18 @@ void RenderPass::DestroyApiObjects()
 
 VkResult CreateTransientRenderPass(
-    VkDevice              device,
+    grfx::Device*         device,
```
Done.

hysw commented:
> but I'll also need a way to set this from a knob.

Adding two boolean flags `KnobFlag<bool>` to `StandardOptions` for each of vrs and fdm should be reasonably simple. If you want force to have only one of them there's always the option of tossing an error.

bjoeris commented:
> Adding two boolean flags `KnobFlag<bool>` to `StandardOptions` for each of vrs and fdm should be reasonably simple. If you want force to have only one of them there's always the option of tossing an error.

Yeah, that worked. The order of `Config` and `InitKnobs` seems confusing, but I think I got it figured out.

I also changed the `ShadingRateCapabilities` for the device to report only the single `ShadingRateMode` that is enabled on the device.

Keenuts commented on include/ppx/application.h:
```diff
@@ -254,6 +254,10 @@ struct StandardOptions
+
+    std::shared_ptr<KnobFlag<bool>> pEnableFDM;
+    std::shared_ptr<KnobFlag<bool>> pEnableVRS;
+    std::shared_ptr<KnobFlag<bool>> pDisableShadingRate;
```
Is this flag really needed? Seems redundant given we have the 2 others to control VRS.

bjoeris replied:
Yeah, I think it makes sense to have all three. The absence of any of these flags indicates to use whatever default the application specified in `ApplicationSettings`, and `--disable-shading-rate` explicitly says to use the `SHADING_RATE_NONE`.

bjoeris replied:
This is superseded by the other conversation about knobs below. `--disable-shading-rate` is now `--shading-rate-mode=none`.

Keenuts reviewed (approved):
LGTM, just nitpicking about flags.
(assuming this doesn't have validation errors so far)

apazylbe reviewed (commented):
Would you be able to add a few lines into the PR description on what changes are included here? Since that will be used as a commit message. For example, it shading rate pattern, flags, using renderpass create 2, etc.
Thanks!

bjoeris commented on src/ppx/application.cpp:
```diff
@@ -1133,6 +1149,36 @@ void Application::UpdateStandardSettings()
         PPX_LOG_WARN("Headless or deterministic mode: disabling ImGui");
     }
+
+    if (mStandardOpts.pEnableFDM->GetValue()) {
```
Yeath, I also tried making a string knob that could be set like `--shading-rate=fdm`, which seems more intuitive/easier to remember how to use, but dealing with the strings made the code a little more complicated.

It seems like the right answer is to have something like an enum type knob, where you could specify the valid strings and which value they mapped to.

I could switch to using strings for now, so that `--shading-rate=fdm` works, and then the internal implementation for that could be switched to an enum knob if that's ever added.

angela28chen replied:
I think that it might be possible to use a dropdown knob for this? Maybe something like the knob `pFullscreenQuadsType` and array of string&Enum `kFullscreenQuadsTypes` used in the graphics pipeline benchmark?

angela28chen replied:
...I just realized there might be a bug in `kFullscreenQuadsTypes` being all noise. But the idea is the same.

angela28chen replied:
You could define a knob for the shading rate and have it use `DropdownEntry<ShadingRateMode>`.

Side note: #387 (fix for the above)

bjoeris replied:
It looks like `KnobDropdown` is designed to allow the value to be changed through the UI, correct?

Shading rate mode cannot be changed after the application starts, so displaying a dropdown with different shading rate modes would be confusing.

With the current Knobs types, he best option I see is to use a `KnobFlag<string>`, with a validator to check that the string is one of the valid options ("" (default), "none", "fdm", "vrs"), and convert the string to `ShadingRateMode` here in `UpdateStandardSettings()`. Does that make sense?

angela28chen replied:
Yes, the current knob class structure is unfortunately confusing for the flag vs other classes division. 

I think `KnobFlag<string>` is a good option, and imo another option is to use `KnobDropdown` and just set the visibility to invisible so it's not shown in the UI. This would also help make the knob description message automatically list the options for you.

bjoeris replied:
Thanks, setting it to invisible would help. I went with a `KnobFlag<string>` because it still feels confusing to use a `KnobDropdown` for something that can't change and shouldn't be rendered as a dropdown. I also noticed that `KnobFlag` can display it's value without allowing the user modify it, which is a good fit here.

bjoeris commented on include/ppx/application.h:
```diff
@@ -307,6 +311,11 @@ struct ApplicationSettings
+            // Enable support for this shading rate mode on the device.
+            // The application must not use FDM or VRS without setting this to
+            // the corresponding shading rate mode.
+            grfx::ShadingRateMode supportShadingRateMode = grfx::SHADING_RATE_NONE;
```
This sets the default which the knobs can override. If the user doesn't specify any of the knob flags, the application-specified default will be used.

I think both an app-specific default and knobs to override that default are important.

For most apps, `NONE` is correct, but not for apps specifically designed to use FDM or VRS.

Being overridable via knobs is also important to be able to switch the shading rate mode without rebuilding the application (and this is what I've been using for the foveation benchmarking).

apazylbe replied:
Ok, fair enough. There is some weirdness right now in that you can specify a default value for the knob for an application which would always override the settings value, but since you are not using knob default it probably doesn't matter. 
Would a specific app error out if the shading rate mode is not supported? Say if the app can only support FDM and VRS is requested.

bjoeris replied:
> Would a specific app error out if the shading rate mode is not supported? Say if the app can only support FDM and VRS is requested.

This flag just enables the shading rate mode on the device, but doesn't attempt to change the render pass to use the shading rate mode.

I think the use of the setting and knob might be more clear with an example from the foveation benchmark app in my [experimental-foveation branch](https://github.com/bjoeris/bigwheels/blob/experimental-foveation/benchmarks/foveation/main.cpp):

- In `FoveationBenchmarkApp::Config()` ([benchmarks/foveation/main.cpp:256](https://github.com/bjoeris/bigwheels/blob/ea601395784ddf21860f2f9d7902fb985699e424/benchmarks/foveation/main.cpp#L256)), the default shading rate mode is set in `ApplicationSettings`.
  - Most applications skip this step, so the default shading rate mode would be `SHADING_RATE_NONE`.
- In `Application::UpdateStandardSettings()` ([src/ppx/application.cpp:1147](https://github.com/bjoeris/bigwheels/blob/ea601395784ddf21860f2f9d7902fb985699e424/src/ppx/application.cpp#L1147)), the default shading rate mode is overwritten if the `--shading-rate-mode` flag is used.
- In `Application::InitializeGrfxDevice()` ([src/ppx/application.cpp:354](https://github.com/bjoeris/bigwheels/blob/ea601395784ddf21860f2f9d7902fb985699e424/src/ppx/application.cpp#L354)), the default or overwritten shading rate mode is passed to the `DeviceCreateInfo`. This will cause an error if the device does not support this shading rate mode, or is using the DX12 backend.
- In `vk::Device::ConfigureShadingRateCapabilities()` ([src/ppx/grfx/vk/vk_device.cpp:320](https://github.com/bjoeris/bigwheels/blob/ea601395784ddf21860f2f9d7902fb985699e424/src/ppx/grfx/vk/vk_device.cpp#L320)), the shading rate mode passed to the `DeviceCreateInfo` is saved in the `ShadingRateCapabilities` for the device.
- In `FoveationBenchmarkApp::Setup()` ([benchmarks/foveation/main.cpp:524](https://github.com/bjoeris/bigwheels/blob/ea601395784ddf21860f2f9d7902fb985699e424/benchmarks/foveation/main.cpp#L524)), the application uses the shading rate mode from the device's `ShadingRateCapabilities` to configure the render pass and pipeline.
  - Most applications skip this step, so their render pass wont actually use VRS or FDM, even if it is enabled on the device.

apazylbe replied:
sg, thanks for the detailed walkthrough!

bjoeris commented:
> Would you be able to add a few lines into the PR description on what changes are included here? Since that will be used as a commit message. For example, it shading rate pattern, flags, using renderpass create 2, etc. Thanks!

Do you mean PRs are squashed to a single commit when merged in this repo? I usually leave separate commits, since that gives more specific information to someone searching through the git history. But I'm happy to do whatever works for folks working in this repo.

I don't see how to add extra lines to the PR description. I only see how to edit the title, which is just a single line.

apazylbe commented:
> Do you mean PRs are squashed to a single commit when merged in this repo? I usually leave separate commits, since that gives more specific information to someone searching through the git history. But I'm happy to do whatever works for folks working in this repo.
> 
> I don't see how to add extra lines to the PR description. I only see how to edit the title, which is just a single line.

Yeah, what we usually do is 1 PR = 1 commit. Since we add commits for review feedback, and that way you can in theory sync to a specific commit and it should be valid. Also presumably if we need multiple commits, then maybe it could have been multiple PRs too? Anyway, you should be able to click "..." in the current PR description to edit it.

![image](https://github.com/google/bigwheels/assets/39095209/3729b54a-9a5b-43a6-9471-a4911b11e80a)

bjoeris commented:
> Anyway, you should be able to click "..." in the current PR description to edit it.

I see, I didn't realize first comment = description. Done.

apazylbe reviewed (approved):
Thanks!

bjoeris commented on include/ppx/grfx/grfx_shading_rate.h:
```diff
@@ -0,0 +1,108 @@
+
+    struct
+    {
+        // Minumum/maximum size of the region of the render target
```
Fixed.
